server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/lightningbitcoin/lnurl/LnurlRestService.java

```diff
@@ -0,0 +1,214 @@
+import com.generalbytes.batm.server.extensions.ITransactionRequest;
+import com.generalbytes.batm.server.extensions.IWallet;
+import com.generalbytes.batm.server.extensions.exceptions.UpdateException;
+import com.generalbytes.batm.server.extensions.extra.lightningbitcoin.LightningBitcoinExtension;
+import com.generalbytes.batm.server.extensions.extra.lightningbitcoin.lnurl.dto.ErrorLnurlResponse;
+import com.generalbytes.batm.server.extensions.extra.lightningbitcoin.lnurl.dto.OkLnurlResponse;
+import com.generalbytes.batm.server.extensions.extra.lightningbitcoin.lnurl.dto.WithdrawLnurlResponse;
+import com.google.common.base.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+import java.math.BigDecimal;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+@Path("/")
+@Produces(MediaType.APPLICATION_JSON)
+public class LnurlRestService implements IRestService {
+    private static final Logger log = LoggerFactory.getLogger(LnurlRestService.class);
+    private static final String LBTC = CryptoCurrency.LBTC.getCode();
+    private static final long WITHDRAW_EXPIRY_HOURS = 3 * 24;
+
+    private final LnInvoiceUtil lnInvoiceUtil = new LnInvoiceUtil();
+    private final LnurlUtil lnurlUtil = new LnurlUtil();
+    private final IExtensionContext extensionContext = LightningBitcoinExtension.getExtensionContext();
+
+    @Override
+    public String getPrefixPath() {
+        return "lnurl";
+    }
+
+    @Override
+    public Class getImplementation() {
+        return LnurlRestService.class;
+    }
+
+    @GET
+    public Object ping() {
+        return "BATM LNURL REST Service";
+    }
+
+    /**
+     * This endpoint is called by the user's wallet first, right after scanning the LNURL QR code.
+     * It just sends the data from the request back to the client.
+     * We encode everything that's needed in the QR code to save one DB query here.
+     * The DB query and security checks are done in the confirm endpoint.
+     * <p>
+     * After getting a response from this endpoint the wallet will display
+     * a screen where the user can confirm the withdraw amount (millisats).
+     * <p>
+     * When the user confirms, the wallet will generate a LN invoice
+     * for the given amount of millisats and send it to the callback URL.
+     *
+     * @return an object containing:
+     * - callback URL where to send an invoice to (the withdraw-confirm endpoint)
+     * - withdrawable amount minimum and maximum (which are equal in our case - the user must withdraw full transaction amount)
+     * - description that *could* be displayed to the user in their wallet
+     */
+    @GET
+    @Path(LnurlUtil.WITHDRAW_PATH)
+    public Object withdraw(@QueryParam("uuid") String uuid, @QueryParam("rid") String rid, @QueryParam("millisats") Long millisats) {
```

Comment:
Instead of returning object it would be better to return some wrapper object used as identification of all response objects (WithdrawLnurlResponse, OkLnurlResponse, ErrorLnurlResponse) which should extend from it.

Reply:
I tried, spend a lot of time on it and I give up :) I cannot make it work, it does not serialize the fields from the subclasses when it extends a class or implements an interface...